### PR TITLE
parquet bloom filter part III: add sbbf writer, remove `bloom` default feature, add reader properties

### PR DESCRIPTION
**Pull request by Jimexist**
# Which issue does this PR close?

- previous PR: #3102
- related issue: #3023 

# Rationale for this change
 
- add sbbf filter during metadata writing phase
- add reading support for bloom filter in reader properties
- create bloom filter per column chunk per row group as per writer properties
- remove `bloom` feature

# What changes are included in this PR?

# Are there any user-facing changes?

now the API is considered _complete_, the next step is to have an end to end test or cross test with actual parquet file generation.

### CODE REVIEW

File: parquet/Cargo.toml [hunk 1]
```diff
@@ -57,7 +57,8 @@ seq-macro = { version = "0.3", default-features = false }
 futures = { version = "0.3", default-features = false, features = ["std"], optional = true }
 tokio = { version = "1.0", optional = true, default-features = false, features = ["macros", "rt", "io-util"] }
 hashbrown = { version = "0.13", default-features = false }
-twox-hash = { version = "1.6", optional = true }
+twox-hash = { version = "1.6", default-features = false }
```
Comment (Jimexist) [inline parquet/Cargo.toml:60]: it by default relies on `rand` which breaks `wasm32`

**Comment by alamb** [comment]
I believe @tustvold  is away for a few days. I plan to review this PR in more detail tomorrow

File: parquet/Cargo.toml [hunk 2]
```diff
@@ -77,7 +78,7 @@ rand = { version = "0.8", default-features = false, features = ["std", "std_rng"
 all-features = true
 
 [features]
-default = ["arrow", "bloom", "snap", "brotli", "flate2", "lz4", "zstd", "base64"]
```
Comment (alamb) [inline parquet/Cargo.toml:80]: 👍

File: parquet/src/bloom_filter/mod.rs [hunk 3]
```diff
@@ -128,6 +172,33 @@ impl Sbbf {
         Self(data)
     }
 
+    /// Write the bitset in serialized form to the writer.
+    pub fn write_bitset<W: Write>(&self, mut writer: W) -> Result<(), ParquetError> {
```
Comment (alamb) [inline parquet/src/bloom_filter/mod.rs:176]: I think it would be good to write a test for round tripping the bloom filters (as in write a SBFF to a Vec<u8> and then read it back out and verify it is the same). Specifically it would be nice to verify the bytes are not scrambled and the lengths are correct and handle empty bitsets (if that is possible)

File: parquet/src/bloom_filter/mod.rs [hunk 4]
```diff
@@ -113,7 +116,48 @@ fn read_bloom_filter_header_and_length(
...
+// where k = number of hash functions, m = number of bits, n = number of distinct values
+#[inline]
+fn num_of_bits_from_ndv_fpp(ndv: u64, fpp: f64) -> f64 {
+    -8.0 * ndv as f64 / (1.0 - fpp.powf(1.0 / 8.0)).ln()
+}
+
 impl Sbbf {
+    /// Create a new [Sbbf] with given number of distinct values and false positive probability.
+    /// Will panic if `fpp` is greater than 1.0 or less than 0.0.
+    pub fn new_with_ndv_fpp(ndv: u64, fpp: f64) -> Self {
```
Comment (alamb) [inline parquet/src/bloom_filter/mod.rs:141]: Since this is a function meant for use outside the parquet crate I would prefer it return an error rather than panic with bad input.

File: parquet/src/bloom_filter/mod.rs [hunk 5]
```diff
@@ -292,4 +363,46 @@ mod tests {
...
+            (0.001, 1000, 14607),
+            (0.1, 10000, 57725),
+            (0.01, 10000, 96815),
+            (0.001, 10000, 146076),
+            (0.1, 100000, 577254),
+            (0.01, 100000, 968152),
+            (0.001, 100000, 1460769),
+            (0.1, 1000000, 5772541),
+            (0.01, 1000000, 9681526),
+            (0.001, 1000000, 14607697),
```
Comment (alamb) [inline parquet/src/bloom_filter/mod.rs:403]: does this mean a 14MB bloom filter? Its ok as there is a limit in `optimal_num_of_bytes ` but when I saw this it was just 🤯 

It might also be good to pass in some value larger than 2^32 to test there isn't an overflow problem lurking

File: parquet/src/file/metadata.rs [hunk 6]
```diff
@@ -236,7 +236,7 @@ pub struct RowGroupMetaData {
 }
 
 impl RowGroupMetaData {
-    /// Returns builer for row group metadata.
+    /// Returns builder for row group metadata.
```
Comment (alamb) [inline parquet/src/file/metadata.rs:239]: 👍

File: parquet/src/file/properties.rs [hunk 7]
```diff
@@ -82,6 +83,9 @@ const DEFAULT_STATISTICS_ENABLED: EnabledStatistics = EnabledStatistics::Page;
 const DEFAULT_MAX_STATISTICS_SIZE: usize = 4096;
 const DEFAULT_MAX_ROW_GROUP_SIZE: usize = 1024 * 1024;
 const DEFAULT_CREATED_BY: &str = env!("PARQUET_CREATED_BY");
+const DEFAULT_BLOOM_FILTER_ENABLED: bool = false;
+const DEFAULT_BLOOM_FILTER_MAX_BYTES: u32 = 1024 * 1024;
```
Comment (alamb) [inline parquet/src/file/properties.rs:87]: 👍

File: parquet/src/file/properties.rs [hunk 8]
```diff
@@ -255,6 +279,11 @@ impl WriterProperties {
             .or_else(|| self.default_column_properties.max_statistics_size())
             .unwrap_or(DEFAULT_MAX_STATISTICS_SIZE)
     }
+
+    def_col_property_getter!(bloom_filter_enabled, bool, DEFAULT_BLOOM_FILTER_ENABLED);
```
Comment (alamb) [inline parquet/src/file/properties.rs:283]: I think these properties need docstrings -- I am happy to help write them. In particular, I think it should mention that the ndv and fpp are knobs that allow for control over bloom filter accuracy. Also it should mention if these limits are for each column chunk or the entire file (e.g. the ndv value is not the number of distinct values in the entire column)

File: parquet/src/bloom_filter/mod.rs [hunk 9]
```diff
@@ -113,7 +116,48 @@ fn read_bloom_filter_header_and_length(
...
+const BITSET_MAX_LENGTH: usize = 128 * 1024 * 1024;
+
+#[inline]
+fn optimal_num_of_bytes(num_bytes: usize) -> usize {
+    let num_bytes = num_bytes.min(BITSET_MAX_LENGTH);
+    let num_bytes = num_bytes.max(BITSET_MIN_LENGTH);
+    num_bytes.next_power_of_two()
+}
+
+// see http://algo2.iti.kit.edu/documents/cacheefficientbloomfilters-jea.pdf
```
Comment (alamb) [inline parquet/src/bloom_filter/mod.rs:129]: Here is the parquet-mr code: https://github.com/apache/parquet-mr/blob/d057b39d93014fe40f5067ee4a33621e65c91552/parquet-column/src/main/java/org/apache/parquet/column/values/bloomfilter/BlockSplitBloomFilter.java#L277-L304

That looks very similar

File: parquet/src/file/properties.rs [hunk 10]
```diff
@@ -599,17 +681,25 @@ impl ColumnProperties {
...
+    def_opt_field_getter!(bloom_filter_enabled, bool);
+    def_opt_field_getter!(bloom_filter_fpp, f64);
+    def_opt_field_getter!(bloom_filter_max_bytes, u32);
+    def_opt_field_getter!(bloom_filter_ndv, u64);
 }
 
 /// Reference counted reader properties.
 pub type ReaderPropertiesPtr = Arc<ReaderProperties>;
 
+const DEFAULT_READ_BLOOM_FILTER: bool = true;
```
Comment (alamb) [inline parquet/src/file/properties.rs:694]: I think this perhaps should default to `false` to maintain the current behavior and avoid (potentially) costly semi-random IO if the reader is not going to read the bloom filter

File: parquet/src/file/reader.rs [hunk 11]
```diff
@@ -145,9 +144,8 @@ pub trait RowGroupReader: Send + Sync {
         Ok(col_reader)
     }
 
-    #[cfg(feature = "bloom")]
     /// Get bloom filter for the `i`th column chunk, if present.
```
Comment (alamb) [inline parquet/src/file/reader.rs:147]:
```suggestion
    /// Get bloom filter for the `i`th column chunk, if present and the reader was configured
    /// to read bloom filters.
```

File: parquet/src/file/writer.rs [hunk 12]
```diff
@@ -212,6 +220,35 @@ impl<W: Write> SerializedFileWriter<W> {
...
 
+    /// Serialize all the bloom filter to the file
+    fn write_bloom_filters(&mut self, row_groups: &mut [RowGroup]) -> Result<()> {
+        // iter row group
+        // iter each column
+        // write bloom filter to the file
+        for (row_group_idx, row_group) in row_groups.iter_mut().enumerate() {
+            for (column_idx, column_chunk) in row_group.columns.iter_mut().enumerate() {
+                match &self.bloom_filters[row_group_idx][column_idx] {
+                    Some(bloom_filter) => {
```
Comment (alamb) [inline parquet/src/file/writer.rs:231]: Given the interactionwith TCompact protocol for reading is in `Sbff` what would you think about moving the interaction for writing there as well (like `Sbff::write(writer)`)?

**Review by alamb — COMMENTED** [review]
Thank you @Jimexist  -- this is very cool. I went through the code fairly thoroughly. I had some minor suggestions / comments for documentation and code structure but nothing that would block merging.

I think the biggest thing I would like to discuss is "what parameters to expose for the writer API". I was thinking, for example, will users of this feature be able to set "fpp" and "ndv" reasonably? I suppose having the number of distinct values before writing a parquet file is reasonable, but maybe not the expected number of distinct values for each row group.

I did some research of other implementations. Here are the spark settingss https://spark.apache.org/docs/latest/configuration.html

spark.sql.optimizer.runtime.bloomFilter.creationSideThreshold | 10MB | Size threshold of the bloom filter creation side plan. Estimated size needs to be under this value to try to inject bloom filter. | 3.3.0
-- | -- | -- | --
spark.sql.optimizer.runtime.bloomFilter.enabled | false | When true and if one side of a shuffle join has a selective predicate, we attempt to insert a bloom filter in the other side to reduce the amount of shuffle data. | 3.3.0
spark.sql.optimizer.runtime.bloomFilter.expectedNumItems | 1000000 | The default number of expected items for the runtime bloomfilter | 3.3.0
spark.sql.optimizer.runtime.bloomFilter.maxNumBits | 67108864 | The max number of bits to use for the runtime bloom filter | 3.3.0
spark.sql.optimizer.runtime.bloomFilter.maxNumItems | 4000000 | The max allowed number of expected items for the runtime bloom filter | 3.3.0
spark.sql.optimizer.runtime.bloomFilter.numBits | 8388608 | The default number of bits to use for the runtime bloom filter | 3.3.0


the arrow parquet C++ writer seems to allow for the fpp setting

https://arrow.apache.org/docs/cpp/api/formats.html#_CPPv4N5arrow8adapters3orc12WriteOptions16bloom_filter_fppE

```
double bloom_filter_fpp = 0.05
The upper limit of the false-positive rate of the bloom filter, default 0.05.
```

Databricks seems to expose the fpp, max_fpp, and num distinct values:
https://docs.databricks.com/sql/language-manual/delta-create-bloomfilter-index.html

**Comment by Jimexist** [comment]
thanks @alamb for the detailed comment. i wish to merge as is and then for subsequent steps:

1. add an integration or round trip tests or maybe both
2. adjust overall configurable parameters and then add docstring after they are settled

do you think this is a good idea given that we are not releasing a new version very soon?

File: parquet/src/file/properties.rs (hunk 10 above)
Comment (Jimexist) [inline parquet/src/file/properties.rs:694]:
```suggestion
const DEFAULT_READ_BLOOM_FILTER: bool = false;
```

Comment (Jimexist) [reply parquet/src/file/properties.rs]: changed to false by default

File: parquet/src/bloom_filter/mod.rs (hunk 4 above)
Comment (Jimexist) [reply parquet/src/bloom_filter/mod.rs:144]: will update once we decided on using fpp or ndv or both:
- #3138

File: parquet/src/file/properties.rs [hunk 13]
```diff
@@ -272,6 +301,52 @@ pub struct WriterPropertiesBuilder {
     sorting_columns: Option<Vec<SortingColumn>>,
 }
 
+macro_rules! def_opt_field_setter {
```
Comment (tustvold) [inline parquet/src/file/properties.rs:304]: Just an observation that these macros are potentially more verbose than the alternative. Perhaps I'm old-fashioned but I'm not a massive fan of macros aside from where absolutely necessary, as they complicate debugging and legibility

File: parquet/src/file/properties.rs [hunk 14]
```diff
@@ -272,6 +301,52 @@ pub struct WriterPropertiesBuilder {
...
+            pub fn [<set_ $field>](&mut self, value: $type) -> &mut Self {
+                self.$field = Some(value);
+                self
+            }
+        }
+    };
+    ($field: ident, $type: ty, $min_value:expr, $max_value:expr) => {
+        paste! {
+            pub fn [<set_ $field>](&mut self, value: $type) -> &mut Self {
+                if ($min_value..=$max_value).contains(&value) {
```
Comment (tustvold) [inline parquet/src/file/properties.rs:316]: I would expect this to error or panic, not just ignore a value out of range?

File: parquet/src/bloom_filter/mod.rs [hunk 15]
```diff
@@ -113,7 +118,43 @@ fn read_bloom_filter_header_and_length(
...
+// where k = number of hash functions, m = number of bits, n = number of distinct values
+#[inline]
+fn num_of_bits_from_ndv_fpp(ndv: u64, fpp: f64) -> usize {
+    let num_bits = -8.0 * ndv as f64 / (1.0 - fpp.powf(1.0 / 8.0)).ln();
+    num_bits as usize
+}
+
 impl Sbbf {
+    /// Create a new [Sbbf] with given number of distinct values and false positive probability.
+    /// Will panic if `fpp` is greater than 1.0 or less than 0.0.
```
Comment (tustvold) [inline parquet/src/bloom_filter/mod.rs:143]:
```suggestion
    /// Will panic if `fpp` is greater than or equal to 1.0 or less than 0.0.
```

File: parquet/src/file/properties.rs [hunk 16]
```diff
@@ -272,6 +301,52 @@ pub struct WriterPropertiesBuilder {
...
+macro_rules! def_opt_field_setter {
+    ($field: ident, $type: ty) => {
+        paste! {
+            pub fn [<set_ $field>](&mut self, value: $type) -> &mut Self {
+                self.$field = Some(value);
+                self
+            }
+        }
+    };
+    ($field: ident, $type: ty, $min_value:expr, $max_value:expr) => {
```
Comment (tustvold) [inline parquet/src/file/properties.rs:313]: This variant it only used in one place, and so I wonder if it needs to be a macro

**Review by tustvold — APPROVED** [review]
I'm happy for this to go in as is, although I personally would prefer to reduce the amount of macros as they make the code quite hard to follow

File: parquet/src/column/writer/mod.rs [hunk 17]
```diff
@@ -231,6 +238,19 @@ impl<'a, E: ColumnValueEncoder> GenericColumnWriter<'a, E> {
         // Used for level information
         encodings.insert(Encoding::RLE);
 
+        let bloom_filter_enabled = props.bloom_filter_enabled(descr.path());
+        let bloom_filter = if bloom_filter_enabled {
+            if let Some(ndv) = props.bloom_filter_ndv(descr.path()) {
+                let fpp = props.bloom_filter_fpp(descr.path());
+                Some(Sbbf::new_with_ndv_fpp(ndv, fpp))
+            } else {
```
Comment (tustvold) [inline parquet/src/column/writer/mod.rs:246]: I think it is perhaps a little surprising that `bloom_filter_max_bytes` is ignored if `ndv` is set

File: parquet/src/file/properties.rs [hunk 18]
```diff
@@ -515,27 +595,24 @@ impl Default for EnabledStatistics {
...
     statistics_enabled: Option<EnabledStatistics>,
     max_statistics_size: Option<usize>,
+    /// bloom filter enabled
+    bloom_filter_enabled: Option<bool>,
+    /// bloom filter expected number of distinct values
+    bloom_filter_ndv: Option<u64>,
+    /// bloom filter false positive probability
+    bloom_filter_fpp: Option<f64>,
+    /// bloom filter max number of bytes
+    bloom_filter_max_bytes: Option<u32>,
```
Comment (tustvold) [inline parquet/src/file/properties.rs:612]: I wonder if it would be simpler to just ask users to specify the bloom filter size, and provide a free function to compute the size based on ndv and fpp?

The interaction of these three properties isn't immediately apparent?

File: parquet/src/file/properties.rs [hunk 19]
```diff
@@ -622,11 +712,17 @@ impl ReaderProperties {
...
+    /// Returns whether to read bloom filter
+    pub(crate) fn read_bloom_filter(&self) -> bool {
+        self.read_bloom_filter
+    }
 }
 
 /// Reader properties builder.
 pub struct ReaderPropertiesBuilder {
     codec_options_builder: CodecOptionsBuilder,
+    read_bloom_filter: Option<bool>,
```
Comment (tustvold) [inline parquet/src/file/properties.rs:725]:
```suggestion
    read_bloom_filter: bool,
```

File: parquet/src/file/properties.rs [hunk 20]
```diff
@@ -635,13 +731,17 @@ impl ReaderPropertiesBuilder {
     fn with_defaults() -> Self {
         Self {
             codec_options_builder: CodecOptionsBuilder::default(),
+            read_bloom_filter: None,
```
Comment (tustvold) [inline parquet/src/file/properties.rs:734]:
```suggestion
            read_bloom_filter: DEFAULT_READ_BLOOM_FILTER,
```

**Comment by Jimexist** [comment]
thanks @tustvold i plan to merge as is and then in subsequent pr adjust:
1. knobs or params that users can tune
2. usage of macro
3. panic behavior

because 2 relies on 1, and 3 relies on 2, so i'd like to clean up 3 of them together

File: parquet/src/file/properties.rs (hunk 18 above)
Comment (alamb) [reply parquet/src/file/properties.rs:612]: That was the conclusion I may have come to as well -- see https://github.com/apache/arrow-rs/issues/3138#issuecomment-1322771730

File: parquet/src/bloom_filter/mod.rs (hunk 9 above)
Comment (XinyuZeng) [reply parquet/src/bloom_filter/mod.rs:131]: @Jimexist Wondering why here we round num_bytes to next power of two? parquet-mr just rounds to (k * BITS_PER_BLOCK)

**Comment by XinyuZeng** [comment]
> the arrow parquet C++ writer seems to allow for the fpp setting
> 
> https://arrow.apache.org/docs/cpp/api/formats.html#_CPPv4N5arrow8adapters3orc12WriteOptions16bloom_filter_fppE

nit: this reference is for arrow ORC C++ writer, parquet C++ does not enable write bloom filter yet.
